### PR TITLE
Stats push interval configurable + linter fixes + improved readability

### PR DESCRIPTION
**Pull request by tchakabam**
This is mostly about getting this into a production-ready shape:

* Debug and min builds
* Linter integration with eslint standard style
* Applying actual linter rules to code
* Some syntactic things to make it more maintainable and readable
* Added a `shakaPlayerInstance` property to fulfill clearly the purpose of accessing Shaka API directly
* Fixed error handling (actually prints proper info here now, previous thing was maybe for previous Shaka major version)

Otherwise there is nothing related to functionnality changes in here. So it should not be risky :)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -32,10 +54,12 @@ export default class DashShakaPlayback extends HTML5Video {
   }
 
   play() {
-    !this._player && this._setup()
+    if (!this._player) {
```
Comment (leandromoreira) [inline index.js]: Why you're changing this style here and not on `textTracks` and others?

File: index.js [hunk 2]
```diff
@@ -1,16 +1,35 @@
-import {HTML5Video, Log, Events} from 'clappr'
-import shaka from 'shaka-player'
+const {HTML5Video, Log, Events} = window.Clappr;
+const shaka = window.shaka;
```
Comment (leandromoreira) [inline index.js]: Why is this better than just relying on webpack / module?

File: index.js [hunk 3]
```diff
@@ -49,40 +73,55 @@ export default class DashShakaPlayback extends HTML5Video {
...
+  get textTracks() { 
+    return this._player && this._player.getTracks().filter((t) => t.type === 'text');
+  }
+  
+  get audioTracks() { 
+    return this._player && this._player.getTracks().filter((t) => t.type === 'audio');
+  }
+  
+  get videoTracks() {
+    return this._player && this._player.getTracks().filter((t) => t.type === 'video');
```
Comment (leandromoreira) [inline index.js]: 👍

File: index.js [hunk 4]
```diff
@@ -1,16 +1,35 @@
...
-const SHAKA_READY = 'shaka:ready'
+const SEND_STATS_AT = 30 * 1000;
+const AUTO = -1;
 
-export default class DashShakaPlayback extends HTML5Video {
-  get name() {return 'dash_shaka_playback'}
+class DashShakaPlayback extends HTML5Video {
 
-  get shakaVersion() {return shaka.player.Player.version}
+  static get Events() {
```
Comment (leandromoreira) [inline index.js]: 👍

File: index.js [hunk 5]
```diff
@@ -49,40 +73,55 @@ export default class DashShakaPlayback extends HTML5Video {
...
-  get videoTracks() {return this._player && this._player.getTracks().filter((t) => t.type === 'video')}
-  getPlaybackType() {return (this._player && this._player.isLive()?'live':'vod') || ''}
+  getPlaybackType() {
+    return (this._player && this._player.isLive() ? 'live' : 'vod') || '';
+  }
 
   selectTrack(track) {
-    this._player.selectTrack(track)
-    this._onAdaptation()
+    this._player.selectTrack(track);
```
Comment (leandromoreira) [inline index.js]: we should either use ; or not use them, do you agree?

Comment (tchakabam) [inline index.js]: yes. 

eslint should be added

File: index.js (hunk 1 above)
Comment (tchakabam) [inline index.js]: yeah good point.

i guess it was just a quick walk through of things i though were ugly and wouldn't pass my linter rules.

File: index.js (hunk 2 above)
Comment (tchakabam) [inline index.js]: yes that was a quick fix to make my build go through.

i agree modules is better and then eventually telling your toolchain to take that as external dep.

Comment (tchakabam) [inline index.js]: will revert that part.

File: index.js (hunk 5 above)
Comment (leandromoreira) [inline index.js]: 👍

**Comment by leandromoreira** [comment]
great, I'll wait your modifications

**Comment by tchakabam** [comment]
@leandromoreira fixed the most functionnaly important part, the webpack externals deps, and also added some linter config and an additional property to get the shaka Player instance.

**Comment by leandromoreira** [comment]
can you also solve the conflicts?

**Comment by tchakabam** [comment]
OK thanks for your support, I put some additional info about the purpose of this into the description.

**Comment by tchakabam** [comment]
What would be great for our integration would be to make the stats interval configurable ... Do you have a suggestion how to do that? I'd be ready to implement this here..

**Comment by leandromoreira** [comment]
Maybe we should think about this stats to be integrated at https://github.com/clappr/clappr-stats

**Comment by tchakabam** [comment]
oh yeah great I had not seen this existed too :) 

yes that should be the case. anyway, they should be posted more often than just every 30 seconds if you wish so.

I have added a config option `shakaSendStatsInterval` in my last commit here, which would be read from to overload the defaults in the plugin here.

we have test monitoring of dropped frames on test devices and it needs to display directly or are using dropped frames stats for our ABR for example, and we need to react more quickly there :)

**Comment by leandromoreira** [comment]
@tchakabam can you tell where you're using clappr? if yes, please go to https://github.com/clappr/clappr/issues/522 and leave it there. :)

**Comment by tchakabam** [comment]
@leandromoreira so would it be good to merge this then?

**Comment by leandromoreira** [comment]
@tchakabam thank you but my main problem is to merge without second thoughts, for instance, I dislike the code format like `function ()` with spaces but I'll ask @towerz

**Comment by tchakabam** [comment]
hey yeah i understand sure.

i have changed the title to reflect better what was done, and also because it was a crappy title i put first :)

furthermore, we realized that this plugin isn't actually future-proof when it comes to Shaka API, as some things have been deprecated (basically the API to get tracks) and we'll be happy to contribute an update here as well.

**Comment by tchakabam** [comment]
@leandromoreira but i think the changes are functionnaly very safe and marginal, the reason there so many line-changes was mainly to improve readability and linter-rules related

**Comment by tchakabam** [comment]
@leandromoreira OK but if you dislike the linter-rules we can change them of course. i just wanted to make it more readable really and actually apply some rules. i am totally open to the nature of the rules what concerns little details really! i am totally not attached to these details wether there are spaces after `function` or not! :D

**Comment by tchakabam** [comment]
@leandromoreira I update the rules file to how it was before and re-ran the linter --fix command on top. turns out there was just one small semicolon change from it. 

if you dislike the style, i d say we should update the existing linter rules in the project accordingly - currently it is just applying the ones that were there already before this PR i.e the ones currently in the master eslintrc, just that they were not applied before, which seemed kind off inconsistent, which is why i wanted to fix that here.

File: package.json [hunk 6]
```diff
@@ -10,18 +10,23 @@
...
   "author": "Clappr team",
   "license": "BSD-3-Clause",
-  "dependencies": {
-    "shaka-player": "v2.0.9",
-    "clappr": "0.2.66"
-  },
   "devDependencies": {
-    "webpack": "^1.12.1",
-    "webpack-dev-server": "^1.10.1",
+    "shaka-player": "2.0.6",
```
Comment (tchakabam) [inline package.json]: `v2.0.9` doesn't exist in NPM, it doesn't seem to be a valid release actually? 

`2.0.8` is the latest release on the `2.0.x` minor.

File: package.json [hunk 7]
```diff
@@ -10,18 +10,25 @@
...
-    "release": "node_modules/.bin/webpack -p"
+    "release": "node_modules/.bin/webpack",
+    "lint": "node_modules/.bin/eslint index.js"
   },
   "author": "Clappr team",
   "license": "BSD-3-Clause",
   "dependencies": {
-    "shaka-player": "v2.0.9",
-    "clappr": "0.2.66"
+    "shaka-player": "2.0.8",
```
Comment (tchakabam) [inline package.json]: `v2.0.9` doesn't exist in NPM, it doesn't seem to be a valid release actually? 

`2.0.8` is the latest release on the `2.0.x` minor.

**Comment by leandromoreira** [comment]
Thank you so much @tchakabam I agree with you too.

**Comment by leandromoreira** [comment]
@tchakabam ;) thank you, feel free to send more PR, to solve the new api for getting audio/text tracks, before we release a new version.

**Comment by leandromoreira** [comment]
We can also update shaka itself to the latest version 2.1.2 !

File: package.json (hunk 7 above)
Comment (leandromoreira) [inline package.json]: that's weird but you're right!

File: index.js [hunk 8]
```diff
@@ -1,20 +1,42 @@
...
+
+  static canPlay (resource, mimeType = '') {
+    shaka.polyfill.installAll()
+    var browserSupported = shaka.Player.isBrowserSupported()
+    var resourceParts = resource.split('?')[0].match(/.*\.(.*)$/) || []
+    return browserSupported && ((resourceParts[1] === 'mpd') || mimeType.indexOf('application/dash+xml') > -1)
+  }
+
+  get name () {
+    return 'DashShakaPlayback'
```
Comment (vagnervjs) [inline index.js]: Hi guys
I know this PR has already been closed (a great contribution, btw), but I think that changing the plugin name can cause side effects in projects that already use this plugin and hope that this value will not (never) be changed, so this seems like a breaking change.
Also, the standard of _Clappr's_ plugins names is `snake_case`, isn't?
What do you think @leandromoreira @tchakabam?

**Comment by kslimani** [comment]
this *may* not the only breaking change. I think something is wrong in Webpack configuration : if i am not wrong, it compile the plugin in es5 without the shaka player dependency. (but as external dep.).

File: index.js (hunk 8 above)
Comment (tchakabam) [inline index.js]: hey @vagnervjs , is that so? 

in that case we may have to change the name back :)

i was under the impression other plugins had camel-case names in fact. but i just checked and you seem to be right! 

sorry for this mistake. you can do a PR where you change the name. would that be ok? Thanks a lot!

**Comment by tchakabam** [comment]
@kslimani the webpack config and build work fine for me, we are actually using this in prod.

The fact we are not including the Shaka dependency is intended :) 

So you can just ship it like you want in seperate, for example a different (newer) version that in this package.

Or say you d like to run the plugin with a patched/modified version, then I would not like to be stuck with a compiled-into-the-plugin Shaka build here.

In general a plugin build should not contain or statically link the dependencies that it "plugs" to, it should link to them dynamically. That is what makes plugins great imho.

Btw that has been like this before this PR (or at least was intended).

**Comment by kslimani** [comment]
@tchakabam i understand your motivations. I am just pointing out that is it a breaking change from previous versions. (_maybe it should be notified in the readme ?_).

An alternative would be to also compile a "light" version of the plugin which does not include Shaka and still compile the actual plugin with Shaka to not introduce breaking changes ?

**Comment by tchakabam** [comment]
@kslimani That sounds like a really good idea in fact! I like that. Do you want to make a PR for that?

**Comment by kslimani** [comment]
@tchakabam i will try to take a look this week.

File: index.js (hunk 8 above)
Comment (vagnervjs) [inline index.js]: Done #27

**Comment by tchakabam** [comment]
@kslimani awesome! 

I don't think the name "light" would be really best though :)

maybe we could call it `plugin` vs `bundle` ?

**Comment by tchakabam** [comment]
@leandromoreira what is your opinion on this?

**Comment by kslimani** [comment]
Well, the current Playback plugin filename is `dash-shaka-playback.js`; I think the name should be left unchanged and be the version bundled with shaka-player.

For the filename without shaka-player bundled, i am not sure myself, maybe something like `dash-shaka-external-playback.js` ? (_the more important, imho, is to properly explain it in readme_).

**Comment by kslimani** [comment]
I created an issue #28 to continue discution.

**Comment by leandromoreira** [comment]
> Well, the current Playback plugin filename is dash-shaka-playback.js; I think the name should be left unchanged and be the version bundled with shaka-player.
👍 
> For the filename without shaka-player bundled, i am not sure myself, maybe something like dash-shaka-external-playback.js ? (the more important, imho, is to properly explain it in readme).
👍 

@kslimani is an amazing developer 👏  thank you once again!

**Comment by tchakabam** [comment]
Awesome, i'm happy we could all agree then on a solution that works for everybody :)

**Pull request merged**